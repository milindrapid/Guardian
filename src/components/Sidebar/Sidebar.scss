@import "../../stylesheet/color";

ul,
li {
  margin: 0;
  padding: 0;
  list-style: none;
}
li {
  padding: 0.75rem;
  font-size: 1rem;
  font-weight: 500;
}
.sidebar {
  width: 250px;
  min-height: calc(100vh);
  box-shadow: $mid-grey 1px 1px 1px;
  background-color: $primary;
  .sidebar-logo {
    height: auto;
    width: 150px;
  }
  a {
    text-decoration: none;
    color: $white;
    &:active,
    &:hover,
    &:focus {
      color: $white;
      text-decoration: none;
    }
  }
  // & > ul > li {
  // border-bottom: $mid-grey 1px solid;
  // }
  .sidebar__item {
    display: flex;
    & + ul {
      padding-top: 0.75rem;
    }
  }
  .sidebar__expand-btn {
    margin-left: auto;
    &.active {
      color: $primary;
    }
  }
}
