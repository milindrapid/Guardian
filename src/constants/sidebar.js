export const sidebar = {
  "submenu-0": {
    name: "Home",
    position: 50,
    description: null,
    is_expanded: false,
    url_string: "/",
    url_str: {},
  },
  "submenu-1": {
    name: "Tools",
    position: 50,
    description: null,
    is_expanded: false,
    url_string: "/tools",
    url_str: {},
    // child: {},
  },
  "submenu-2": {
    name: "Etc",
    position: 50,
    description: null,
    is_expanded: false,
    url_string: "/",
    url_str: {},
  },
};
