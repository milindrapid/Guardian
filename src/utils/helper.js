// format objects to array items
export const formatToArrayItems = (obj) =>
  Object.keys(obj).map((item) => {
    const childrenObj =
      (obj[item].child && formatToArrayItems(obj[item].child)) || null;
    return { id: item, children: childrenObj, ...obj[item] };
  });
