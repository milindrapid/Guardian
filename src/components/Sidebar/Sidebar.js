import React from "react";
import PropTypes from "prop-types";
import SidebarItem from "./SidebarItem";
import "./Sidebar.scss";
import logo from "../../images/logo.png";

const Sidebar = ({ items, depthStep, depth }) => {
  return (
    <div className="sidebar">
      <img className="sidebar-logo" src={logo} alt="Guardian" />
      <ul>
        {items.map((sidebarItem, index) => (
          <SidebarItem
            key={`${sidebarItem.id}${index}`}
            depthStep={depthStep}
            depth={depth}
            isExpanded={sidebarItem.is_expanded}
            to={sidebarItem.url_string}
            {...sidebarItem}
          />
        ))}
      </ul>
    </div>
  );
};

Sidebar.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  depthStep: PropTypes.number,
  depth: PropTypes.number,
};

Sidebar.defaultProps = {
  depthStep: 10,
  depth: 0,
};

export default Sidebar;
