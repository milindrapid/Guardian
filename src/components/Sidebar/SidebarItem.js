import React, { useState } from "react";
import PropTypes from "prop-types";
import "./Sidebar.scss";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faAngleUp, faAngleDown } from "@fortawesome/free-solid-svg-icons";
import { Link } from "react-router-dom";

const SidebarItem = (props) => {
  const [isExpanded, setIsEcpanded] = useState(props.isExpanded);
  const label = props.name;
  const items = props.children;
  const to = props.to;
  const depthStep = 10;
  const depth = 0;
  return (
    <>
      <li>
        <div
          className="sidebar__item"
          style={{ paddingLeft: depth * depthStep }}
        >
          <Link to={to}>
            <span>{label}</span>
          </Link>
          {Array.isArray(items) && (
            <div
              className="sidebar__expand-btn"
              onClick={() => setIsEcpanded(!isExpanded)}
            >
              <FontAwesomeIcon
                icon={isExpanded ? faAngleUp : faAngleDown}
                className={isExpanded ? "active" : ""}
              />
            </div>
          )}
        </div>
        {isExpanded && (
          <>
            {Array.isArray(items) ? (
              <ul>
                {items.map((subItem) => {
                  return (
                    <SidebarItem
                      key={subItem.name}
                      depth={depth + 1}
                      depthStep={depthStep}
                      isExpanded={subItem.is_expanded}
                      {...subItem}
                    />
                  );
                })}
              </ul>
            ) : null}
          </>
        )}
      </li>
    </>
  );
};

SidebarItem.propTypes = {
  isExpanded: PropTypes.bool,
  name: PropTypes.string,
  to: PropTypes.string,
  children: PropTypes.arrayOf(PropTypes.object),
};

SidebarItem.defaultProps = {
  isExpanded: false,
  name: "",
  to: "/",
  children: [],
};

export default SidebarItem;
