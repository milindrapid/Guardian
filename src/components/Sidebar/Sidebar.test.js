import React from "react";
import { shallow } from "enzyme";
import Sidebar from "./Sidebar";
import { formatToArrayItems } from "../../utils/helper";
import { sidebar } from "../../constants/sidebar";

describe("Sidebar Component", () => {
  const props = {
    items: formatToArrayItems(sidebar),
  };
  const wrapper = shallow(<Sidebar {...props} />);

  it("Sidebar Component should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("Should render children correctly", () => {
    expect(wrapper.find("div")).toHaveLength(1);
    expect(wrapper.find("ul")).toHaveLength(1);
  });
});
