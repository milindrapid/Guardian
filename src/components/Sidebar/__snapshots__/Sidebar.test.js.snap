// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar Component Sidebar Component should render correctly 1`] = `
<div
  className="sidebar"
>
  <img
    alt="Guardian"
    className="sidebar-logo"
    src="logo.png"
  />
  <ul>
    <SidebarItem
      depth={0}
      depthStep={10}
      description={null}
      id="submenu-0"
      isExpanded={false}
      is_expanded={false}
      key="submenu-00"
      name="Home"
      position={50}
      to="/"
      url_str={Object {}}
      url_string="/"
    />
    <SidebarItem
      depth={0}
      depthStep={10}
      description={null}
      id="submenu-1"
      isExpanded={false}
      is_expanded={false}
      key="submenu-11"
      name="Tools"
      position={50}
      to="/tools"
      url_str={Object {}}
      url_string="/tools"
    />
    <SidebarItem
      depth={0}
      depthStep={10}
      description={null}
      id="submenu-2"
      isExpanded={false}
      is_expanded={false}
      key="submenu-22"
      name="Etc"
      position={50}
      to="/"
      url_str={Object {}}
      url_string="/"
    />
  </ul>
</div>
`;
